image.go

```diff
@@ -123,50 +127,88 @@ func (t MediaTypes) LayerType() types.MediaType {
+func OverrideMediaTypes(image v1.Image, mediaTypes MediaTypes) (v1.Image, error) {
 	if mediaTypes == DefaultTypes || mediaTypes == MissingTypes {
 		// without media types option, default to original media types
-		return base, nil
+		return image, nil
 	}
 
 	// manifest media type
-	image := mutate.MediaType(empty.Image, mediaTypes.ManifestType())
-	// update empty image with base config
-	config, err := base.ConfigFile()
+	retImage := mutate.MediaType(empty.Image, mediaTypes.ManifestType())
+
+	// update empty image with image config
+	config, err := image.ConfigFile()
 	if err != nil {
 		return nil, err
 	}
+	history := config.History
+	// zero out diff IDs and history, as these will be updated when we call `mutate.Append`
 	config.RootFS.DiffIDs = make([]v1.Hash, 0)
-	image, err = mutate.ConfigFile(image, config)
+	config.History = []v1.History{}
+	retImage, err = mutate.ConfigFile(retImage, config)
 	if err != nil {
 		return nil, err
 	}
 
 	// config media type
-	image = mutate.ConfigMediaType(image, mediaTypes.ConfigType())
+	retImage = mutate.ConfigMediaType(retImage, mediaTypes.ConfigType())
 
 	// layers media type
-	layers, err := base.Layers()
+	layers, err := image.Layers()
 	if err != nil {
 		return nil, err
 	}
-	additions := layersAddendum(layers, mediaTypes.LayerType())
-	image, err = mutate.Append(image, additions...)
+	additions := layersAddendum(layers, history, mediaTypes.LayerType())
+	retImage, err = mutate.Append(retImage, additions...)
 	if err != nil {
 		return nil, err
 	}
 
-	return image, nil
+	return retImage, nil
+}
+
+// OverrideHistoryIfNeeded zeroes out the history if the number of history entries doesn't match the number of layers.
+func OverrideHistoryIfNeeded(image v1.Image) (v1.Image, error) {
+	configFile, err := image.ConfigFile()
+	if err != nil || configFile == nil {
+		return nil, fmt.Errorf("getting image config: %w", err)
+	}
+	configFile.History = NormalizedHistory(configFile.History, len(configFile.RootFS.DiffIDs))
+	return mutate.ConfigFile(image, configFile)
+}
+
+func NormalizedHistory(history []v1.History, nLayers int) []v1.History {
+	if history == nil {
+		return make([]v1.History, nLayers)
+	}
+	if len(history) == nLayers {
+		return history
+	}
+	// try to remove history for empty layers
+	var nHistory []v1.History
+	for _, h := range history {
+		if !h.EmptyLayer {
+			nHistory = append(nHistory, h)
+		}
+	}
+	if len(nHistory) == nLayers {
+		return nHistory
+	}
+	return make([]v1.History, nLayers)
 }
 
 // layersAddendum creates an Addendum array with the given layers
 // and the desired media type
-func layersAddendum(layers []v1.Layer, mediaType types.MediaType) []mutate.Addendum {
+func layersAddendum(layers []v1.Layer, history []v1.History, mediaType types.MediaType) []mutate.Addendum {
 	additions := make([]mutate.Addendum, 0)
-	for _, layer := range layers {
+	if len(history) != len(layers) {
+		history = make([]v1.History, len(layers))
```

Comment:
Will this lead to an off by N type of association if we get here with partial history? Not super worried about that but curious.

Reply:
I think we should be okay... we "normalize" the history (remove empty layers, check that the lengths are equal) at every stage - on initialization, when adding a layer, on save.